python/lsst/ts/watcher/rules/atcamera_dewar.py

```diff
@@ -0,0 +1,458 @@
+                        for both vacuum and temperatures. (Treating these categories separately
+                        would make it difficult to warn if we never got enough data
+                        for the category that requires more data).
+                        Once the rule first sees this much data, it will issue a warning
+                        whenever the number of data points drops below this value.
+                    type: integer
+                    default: 10
+                    exclusiveMinimum: 0
+                max_data_age:
+                    description:
+                        The maximum time this rule will wait for data (seconds)
+                        before issuing a serious alarm.
+                    type: number
+                    default: 60
+                    exclusiveMinimum: 0
+            required:
+              - temperature_window
+              - temperature_hysteresis
+              - vacuum_window
+              - vacuum_hysteresis
+              - min_values
+              - max_data_age
+            additionalProperties: false
+        """
+        return yaml.safe_load(schema_yaml)
+
+    async def no_data_timer(self):
+        """Timer for no data received."""
+        await asyncio.sleep(self.config.max_data_age)
+        self.alarm.set_severity(
+            severity=AlarmSeverity.SERIOUS,
+            reason=f"No data seen in {self.config.max_data_age} seconds",
+        )
+
+    def reset_all(self):
+        """Reset the alarm, clear the data queue, and restart the no-data
+        timer.
+        """
+        self.alarm.reset()
+        self.data_queue.clear()
+        self.had_enough_data = False
+        self.restart_no_data_timer()
+
+    def restart_no_data_timer(self):
+        """Start or restart the no-data timer."""
+        self.no_data_timer_task.cancel()
+        self.no_data_timer_task = asyncio.create_task(self.no_data_timer())
+
+    def start(self):
+        self.reset_all()
+        self.restart_no_data_timer()
+
+    def stop(self):
+        self.no_data_timer_task.cancel()
+
+    def start_call_counter(self, n):
+        """Start self.call_counter_task, which waits for n calls.
+
+        Intended for unit tests.
+        """
+        if not self.call_counter_task.done():
+            raise RuntimeError("Already running")
+        self.calls_remaining = n
+        self.call_counter_task = asyncio.Future()
+
+    def __call__(self, topic_callback: watcher.TopicCallback) -> AlarmSeverity:
+        try:
+            self.restart_no_data_timer()
+            self.data_queue.append(topic_callback.get())
+            curr_tai = utils.current_tai()
+
+            oldest_temp_tai = curr_tai - self.config.temperature_window
+            oldest_vacuum_tai = curr_tai - self.config.vacuum_window
+            oldest_tai = min(oldest_temp_tai, oldest_vacuum_tai)
+
+            while self.data_queue:
```

Comment:
does this loop need to run this fast? Will it chew through an unnecessary amount of CPU cycles?

Reply:
I'm not sure I understand your objection. `__call__` is called whenever a new `vacuum` topic is read. This adds a new data point to the data queue, and the data is processed. Processing includes:
* Purge old data (that's what happens in this loop).
* Compute medians of what's left.
* Pass the medians to the threshold handler to see which alarm conditions to report, if any.